qiita_db/util.py

```diff
@@ -2673,3 +2678,216 @@ def _resource_allocation_failures(df, k, a, b, model, col_name, type_):
+        # - external_id:  overall container for the job and its associated
+        #                   requests. When the Timelimit is hit, the container
+        #                   would take care of completing/stopping the
+        #                   external_id.batch job.
+        # - external_id.batch: it's a container job, it provides how
+        #                        much memory it uses and cpus allocated, etc.
+        # - external_id.extern: takes into account anything that happens
+        #                       outside processing but yet is included in
+        #                       the container resources. As in, if you ssh
+        #                       to the node and do something additional or run
+        #                       a prolog script, that processing would be under
+        #                       external_id but separate from external_id.batch
+        # Here we are going to merge all this info into a single row + some
+        # other columns
+
+        def merge_rows(rows):
+            date_fmt = '%Y-%m-%dT%H:%M:%S'
+            wait_time = (
+                datetime.strptime(rows.iloc[0]['Start'], date_fmt)
+                - datetime.strptime(rows.iloc[0]['Submit'], date_fmt))
+            tmp = rows.iloc[1].copy()
+            tmp['WaitTime'] = wait_time
+            return tmp
+
+        curr = _d.groupby(
+                'external_id').apply(merge_rows).reset_index(drop=True)
+
+        row_dict = {
+            'processing_job_id': job.id,
+            'external_id': eid,
+            'sId': s.id,
+            'sName': sname,
+            'sVersion': s.version,
+            'cId': cmd.id,
+            'cName': cmd.name,
+            'samples': samples,
+            'columns': columns,
+            'input_size': input_size,
+            'extra_info': extra_info,
+            'ElapsedRaw': curr['ElapsedRaw'].iloc[0],
+            'MaxRSS': curr['MaxRSS'].iloc[0],
+            'Submit': curr['Submit'].iloc[0],
+            'Start': curr['Start'].iloc[0],
+            'WaitTime': curr['WaitTime'].iloc[0],
+            'CPUTimeRAW': curr['CPUTimeRAW'].iloc[0],
+            'ReqMem': curr['ReqMem'].iloc[0],
+            'AllocCPUS': curr['AllocCPUS'].iloc[0],
+            'AveVMSize': curr['AveVMSize'].iloc[0]
+        }
+
+        data.append(row_dict)
+    df = pd.DataFrame(data)
+
+    # This is important as we are transforming the MaxRSS to raw value
+    # so we need to confirm that there is no other suffixes
+    print('Make sure that only 0/K/M exist', set(
+        df.MaxRSS.apply(lambda x: str(x)[-1])))
+
+    # Generating new columns
+    df['MaxRSSRaw'] = df.MaxRSS.apply(lambda x: MaxRSS_helper(str(x)))
+    df['ElapsedRawTime'] = df.ElapsedRaw.apply(
+        lambda x: timedelta(seconds=float(x)))
+
+    for index, row in df.iterrows():
+        with qdb.sql_connection.TRN:
+            sql = """
+                INSERT INTO qiita.slurm_resource_allocations (
```

Comment:
The resources used are sensitive to extrinsic factors like other work on the system or cluster. Scoping to include that type of information carries a reasonable amount of complexity, so I don't think we want to attempt that. However, if the date the job was started on was tracked, then it would at least allow for excluding periods of known high burden that artificially increase resource use.

There are also intrinsic factors though that are important for interpretation of resource use, such as the node which ran the job, and ideally the model of processor (`/proc/cpuinfo`) on the node (as hostnames are not assured to be unique over time). It would be great to have information about the memory of the system but that may require sudo (see e.g., `dmidecode`) so could be out of scope. These values are important as our compute cluster is heterogeneous

Reply:
@wasade, if I understand the comment correctly basically you are suggesting storing: 
a. the node name
b. the submission timestamp
correct?

Reply:
(a) yes (b) the job start time stamp, not submission and ideally (c) the actual CPU model on the node

Reply:
Got it, thank you for the clarification. BTW why explicitly (c) vs. only (a)? Also, will `sacct` return (c)?

Reply:
"... hostnames are not assured to be unique over time..." so just getting hte name isn't reliable over time. I'm not sure what information is or isn't easily available from slurm